### PR TITLE
Fix to prevent stopping buffering prematurely

### LINKED ISSUE #16438: Bug Report: Premature buffering stop during concurrent external reparenting can lead to query failures
### Overview of the Issue

With the `keyspace_events` buffer implementation, we see that sometimes buffering stops before the failover has actually been detected and processed by the `healthcheck` stream, causing buffered queries to be sent to the demoted primary.

Here's the log output from one `vtgate` process:

```
{"_time":"2024-07-22T13:23:30.708+00:00","message":"Starting buffering for shard: <redacted>/20-30 (window: 5s, size: 1000, max failover duration: 5s) (A failover was detected by this seen error: vttablet: rpc error: code = Code(17) desc = The MySQL server is running with the --super-read-only option so it cannot execute this statement (errno 1290) (sqlstate HY000) (CallerID: issues_pull_requests_rw_1).)"}
{"_time":"2024-07-22T13:23:33.840+00:00","message":"Starting buffering for shard: <redacted>/30-40 (window: 5s, size: 1000, max failover duration: 5s) (A failover was detected by this seen error: vttablet: rpc error: code = Code(17) desc = The MySQL server is running with the --super-read-only option so it cannot execute this statement (errno 1290) (sqlstate HY000) (CallerID: issues_pull_requests_rw_1).)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"Adding 1 to PrimaryPromoted counter for target: keyspace:\"<redacted>\" shard:\"20-30\" tablet_type:REPLICA, tablet: <redacted>-0171233832, tabletType: PRIMARY"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/80-90 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/c0-d0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/b0-c0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/e0-f0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/-10 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/f0- resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/60-70 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/40-50 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/d0-e0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/50-60 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/90-a0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/10-20 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/a0-b0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/70-80 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/20-30 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"Stopping buffering for shard: <redacted>/20-30 after: 3.1 seconds due to: a primary promotion has been detected. Draining 50 buffered requests now."}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"disruption in shard <redacted>/30-40 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"Stopping buffering for shard: <redacted>/30-40 after: 0.0 seconds due to: a primary promotion has been detected. Draining 1 buffered requests now."}
{"_time":"2024-07-22T13:23:33.856+00:00","message":"Draining finished for shard: <redacted>/30-40 Took: 195.979µs for: 1 requests."}
{"_time":"2024-07-22T13:23:33.960+00:00","message":"FailoverTooRecent-<redacted>/30-40: NOT starting buffering for shard: <redacted>/30-40 because the last failover which triggered buffering is too recent (104.215357ms < 1m0s). (A failover was detected by this seen error: Code: CLUSTER_EVENT"}
{"_time":"2024-07-22T13:23:34.016+00:00","message":"Draining finished for shard: <redacted>/20-30 Took: 159.349351ms for: 50 requests."}
{"_time":"2024-07-22T13:23:34.642+00:00","message":"not marking healthy primary <redacted>-0171231759 as Up for <redacted>/20-30 because its PrimaryTermStartTime is smaller than the highest known timestamp from previous PRIMARYs <redacted>-0171233832: -62135596800 < 1721654613 "}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"Adding 1 to PrimaryPromoted counter for target: keyspace:\"<redacted>\" shard:\"30-40\" tablet_type:REPLICA, tablet: <redacted>-0171233041, tabletType: PRIMARY"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.743+00:00","message":"keyspace event resolved: <redacted>/<redacted> is now consistent (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/50-60 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/60-70 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/40-50 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/d0-e0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/70-80 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/90-a0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/10-20 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/a0-b0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/30-40 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/20-30 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/f0- resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/80-90 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/c0-d0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/b0-c0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/e0-f0 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:36.744+00:00","message":"disruption in shard <redacted>/-10 resolved (serving: true)"}
{"_time":"2024-07-22T13:23:37.282+00:00","message":"not marking healthy primary <redacted>-0171232808 as Up for <redacted>/30-40 because its PrimaryTermStartTime is smaller than the highest known timestamp from previous PRIMARYs <redacted>-0171233041: -62135596800 < 1721654616 "}
{"_time":"2024-07-22T13:23:38.961+00:00","message":"FailoverTooRecent-<redacted>/30-40: skipped 6 log messages"}
{"_time":"2024-07-22T13:23:38.961+00:00","message":"Execute: skipped 3 log messages"}
```

I think what's happening here is that primaries of the `20-30` and `30-40` shard went into read-only mode due to the external failover at roughly the same time, which in turn caused buffering to start on both these shards in quick succession.

Once the primary failover on shard `20-30` was done and Vitess was notified about the new primary via a `TabletExternallyReparented` call, the whole keyspace was detected as being consistent again - including the `30-40` shard which was still in the midst of an external failover. This caused the buffering on the `20-30` _and_ the `30-40` shard to stop, while the `30-40` shard was not failed over yet.

Queries that performed write operations against the `30-40` shard started noticeably failing, until the external failover was finished.

### Reproduction Steps

N/A

### Binary Version

```sh
v17+
```

### Operating System and Environment details

```sh
N/A
```

### Log Fragments

```sh
N/A
```

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

I'll summarise the issue described in #16438. Please follow all the comments in the issue for more details.

The problem that was noticed was that in the new keyspace events implementation of buffering, when 2 shards were being externally reparented around the same time, the buffering implementation would stop buffering for both shards when the first shard completed reparenting. This fails the writes on the second shard which is still in the midst of reparenting.

I was able to find the underlying problem. Basically, the order of steps are something like this -
1. A vttablet is marked read-only (as part of an external reparent). Any write that comes to it fails with the read-only error and starts buffering.
2. Now let's say a PRS starts for a different shard.
3. The keyspace even watcher sees that a PRS has started, and stores that this shard is non-serving.
4. Later when PRS completes, it sees that the PRS has completed. It however doesn't know that the first vttablet from a different shard than the one that just completed PRS is actually still read-only. So, it just stops buffering!
5. This causes the queries targeted towards the first vttablet to see failures on the application side.

The problem was that when we start buffering, we weren't coordinating with the keyspace event watcher. If it doesn't receive a healthcheck from the primary tablet saying its not-serving, it doesn't even know the shard is in the midst of reparenting.

After realising this, I tried out a change where-in, we mark the shard non-serving in the shard state of the keyspace event watcher when we start buffering. This solution however didn't work, as it created a new issue, wherein we would stop buffering even when the primary tablet that is being demoted sends a serving health check. This is possible during the grace period of PRS, and also during external reparents. 

To mitigate this problem, I've implemented a different solution in this PR. This solution is somewhat similar to how  the previous `healthcheck` implementation used to handle stopping of buffering.

The proposed solution, takes the previous changes a bit further, and when we start buffering after receiving an error that tells us a reparent has started, we ask the keyspace event watcher to only mark the shard serving again after it has seen a serving primary with a higher timestamp than the one it knows off.

This change works well for all the cases we want it to, but has one downside. If a PRS call fails midway, and we end up reverting the primary demotion, we won't be able to stop buffering because we wouldn't have seen a new primary. This is a drawback that existed in the previous implementation of `healthcheck` buffering as well.

However, I have tried to mitigate this somewhat. In our implementation, if we receive a non-serving healthcheck from a primary tablet after buffering has started, we drop the restriction of only needing to see a new higher timestamp primary. This is because, we only use `waitForReparent` to ensure that we don't prematurely stop buffering when we receive a serving healthcheck from the primary that is being demoted. However, if we receive a non-serving check, then we know that we won't receive any more serving healthchecks until reparent finishes. Specifically, this helps us when PRS fails, but stops gracefully because the new candidate couldn't get caught up in time. In this case, we promote the previous primary back. This too will stop buffering since we aren't explicitly waiting for a new primary.

## Related Issue(s)

- Fixes #16438

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go [hunk 1]
```diff
@@ -177,3 +180,66 @@ func TestERSWithWriteInPromoteReplica(t *testing.T) {
...
+			case <-ctx.Done():
+				return
+			case <-tick.C:
+				go func() {
+					conn, err := mysql.Connect(context.Background(), &vtParams)
+					if err != nil {
+						return
+					}
+					// We're running queries every 100 millisecond and verifying the results are all correct.
+					res, err := conn.ExecuteFetch(utils.GetSelectionQuery(), rowCount+10, false)
```
Comment (arthurschreiber) [inline go/test/endtoend/reparent/newfeaturetest/reparent_test.go:224]: So this is only running `SELECT` queries during the failover? `SELECT` queries don't get buffered, and don't fail if the server is put into `read_only` mode during the failover. You'll need to run `DML` queries to actually see any impact. 🤔

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go [hunk 2]
```diff
@@ -177,3 +180,66 @@ func TestERSWithWriteInPromoteReplica(t *testing.T) {
...
+			}
+		}
+	}()
+
+	// Now, we run go routines to run PRS calls on all the shards simultaneously.
+	wg := sync.WaitGroup{}
+	for _, shard := range shards {
+		wg.Add(1)
+		go func() {
+			time.Sleep(time.Second * time.Duration(rand.IntN(6)))
```
Comment (arthurschreiber) [inline go/test/endtoend/reparent/newfeaturetest/reparent_test.go:237]: You might have to remove this `time.Sleep` here. I don't have a good idea of how long the failovers take in the tests, but I'd expect them to take around a second or so, which means it's very unlikely to happen at the same time.

**Comment by arthurschreiber** [comment]
It might be easier to reproduce this via an "external" failover:

* Set up a cluster with 2 shards (`-80`, `80-`)
* Switch `-80` and `80-` primaries into read only mode to start buffering.
* Start running a DML query that's hitting shard `-80` and one DML query that's hitting shard `80-`. Both of these queries should get buffered and "block".
* Perform external promotion of a replica in shard `-80` to primary. Then run `TabletExternallyReparented` for the externally promoted replica to mark it as primary in the vitess topology.
* This will stop buffering for _both_ shards `-80` and `80-`. The query that was buffered for shard `-80` will be retried and will be successful, the query that was buffered for shard `80-` will fail because buffering was ended prematurely.

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go (hunk 1 above)
Comment (GuptaManan100) [reply go/test/endtoend/reparent/newfeaturetest/reparent_test.go:224]: Replica reads aren't buffered, but any query including SELECTs that is target to a PRIMARY tablet are buffered. But yes, the `read_only` wouldn't really fail for SELECTs. I'll try DMLs instead 👍

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go (hunk 2 above)
Comment (GuptaManan100) [reply go/test/endtoend/reparent/newfeaturetest/reparent_test.go:237]: I had tried with and without the sleep. I actually made a few more changes here and there, but nothing seemed to work. I'll try what you suggested in 👇

**Comment by GuptaManan100** [comment]
Thanks Arthur! I was able to reproduce the problem now!

**Comment by GrahamCampbell** [comment]
Should backport labels be applied?

**Comment by arthurschreiber** [comment]
I added backport labels. This is a bug and is quite unexpected and confusing when run into.

**Review by arthurschreiber — APPROVED** [review]
❤️

**Comment by arthurschreiber** [comment]
@GuptaManan100 Is there anything outstanding before this can land? I can try and see if I can give these changes a go in our environments to see if the issues we've run into before are all fixed.

File: go/vt/vtgate/buffer/buffer.go [hunk 3]
```diff
@@ -94,6 +94,18 @@ func CausedByFailover(err error) bool {
 	return isFailover
 }
 
+// IsErrorDueToReparenting is a stronger check than CausedByFailover, meant to return
+// if the failure is caused because of a reparent.
+func IsErrorDueToReparenting(err error) bool {
```
Comment (deepthi) [inline go/vt/vtgate/buffer/buffer.go:99]: This seems to be used only within the `buffer` package. It does not need to be exported.

File: go/vt/discovery/keyspace_events.go [hunk 4]
```diff
@@ -368,8 +372,34 @@ func (kss *keyspaceState) onHealthCheck(th *TabletHealth) {
...
+		case !th.Serving:
+			sstate.serving = false
+		}
+	}
+	if !th.Serving {
+		// Once we have seen a non-serving primary healthcheck, there is no need for us to explicitly wait
+		// for a reparent to happen. We use waitForReparent to ensure that we don't prematurely stop
+		// buffering when we receive a serving healthcheck from the primary that is being demoted.
+		// However, if we receive a non-serving check, then we know that we won't receive any more serving
+		// healthchecks anymore until reparent finishes. Specifically, this helps us when PRS fails, but
```
Comment (deepthi) [inline go/vt/discovery/keyspace_events.go:398]:
```suggestion
		// health checks until reparent finishes. Specifically, this helps us when PRS fails, but
```

**Review by deepthi — APPROVED** [review]

**Pull request merged**